gensim/models/keyedvectors.py

```diff
@@ -336,6 +346,95 @@ def most_similar(self, positive=[], negative=[], topn=10, restrict_vocab=None, i
+        in vocabulary of model, restriced by the supplied restrict_vocab.
+
+        'restrict_vocab' should be a list/set of words. The returned word similarities will only
+        contain similarity scores for those words that are in restrict_vocab (and in trained vocabulary).
+
+        If some words in restrict_vocab are not in vocabulary then a warning is issued to the user.
+        This can be supressed by setting the suppress_warning flag.
+
+        Example::
+
+          >>> trained_model.most_similar_among(positive=['man'], topn=1, restrict_vocab=['woman','random_word'])
+          [('woman', 0.75882536)]
+
+        """
+        self.init_sims()
+
+        if isinstance(positive, string_types) and not negative:
+            # allow calls like most_similar('dog'), as a shorthand for most_similar(['dog'])
+            positive = [positive]
+
+        # add weights for each word, if not already present; default to 1.0 for positive and -1.0 for negative words
+        positive = [
+            (word, 1.0) if isinstance(word, string_types + (ndarray,)) else word
+            for word in positive
+        ]
+        negative = [
+            (word, -1.0) if isinstance(word, string_types + (ndarray,)) else word
+            for word in negative
+        ]
+
+        # compute the weighted average of all words
+        all_words, mean = set(), []
+        for word, weight in positive + negative:
+            if isinstance(word, ndarray):
+                mean.append(weight * word)
+            else:
+                mean.append(weight * self.word_vec(word, use_norm=True))
+                if word in self.vocab:
+                    all_words.add(self.vocab[word].index)
+        if not mean:
+            raise ValueError("cannot compute similarity with no input")
+        mean = matutils.unitvec(array(mean).mean(axis=0)).astype(REAL)
+
+        if indexer is not None:
+            return indexer.most_similar(mean, topn)
+
+        if isinstance(restrict_vocab, int):
+            raise ValueError("restrict_vocab must be a set/list of words. Maybe you wanted the most_similar function.")
+        elif isinstance(restrict_vocab, list) or isinstance(restrict_vocab, set):
+            restrict_vocab = set(restrict_vocab)
+            vocabulary_words = set(self.get_words_from_vocab())
+
+            words_to_use = vocabulary_words.intersection(restrict_vocab)
+
+            if(not suppress_warning):
+                toWarn = "The following words are not in trained vocabulary : "
+                toWarn += str(restrict_vocab.difference(vocabulary_words))
+                warnings.warn(toWarn, UserWarning)
+            if(not words_to_use):
+                raise ValueError("None of the words in restrict_vocab exist in current vocabulary")
+
+            restrict_vocab_indices = [self.vocab[word].index for word in words_to_use]
+            limited = self.syn0norm[restrict_vocab_indices] #syn0norm is an ndarray
```

Comment:
I think this creates a new array (rather than just a view), which might be a significant memory cost, for large models and when the words-of-interest are a large subset of all words. Maybe this is unavoidable, but might there be any way to do the calculation against the indexes in the original  `syn0norm`?